### PR TITLE
Files Changed Overview ( from #5626 rebased and within scope) 

### PR DESCRIPTION
**Pull request by playcations**
This is a rebased version of PR #5626 to resolve merge conflicts and scope.
 
## Original PR: #5626
**Original Title:** Files Changed Overview (FCO)

## What was done:
* Branched off the head of main and refactored FCO to match  
* It was too far back in the commits it was causing issues with the rebasing
* Removed checkpoint changes as they are outside the scope of this PR
* Will now only work if checkpoints are working. 

Closes Issue: #4454 when merged.
Closes PR: #5626 when merged.

### Related GitHub Issue

Closes: # [4454](https://github.com/RooCodeInc/Roo-Code/issues/4454)

### Description

This adds a collapsable window in chat to show files that have been modified by agents. 

Features include accepting changes, diffing and rejecting file changes outright. 
This prevents you from needing to keep git open or going back to a previous checkpoint after every big modifications to the codebase. I definitely had times where some good changes have happened but then the agent decides theya ren't done! You can easily reject some of the changes instead of all of them. If you don't like this feature, you can also turn it off in the settings menu. 

Checkpoints system did need modifications to get this to work properly. You can see what was done there and let me know what you think. 
ShadowCheckpoints are now stored in a .roo folder in root. This prevents the need to search for all git files and turn off the system if some are found not in root. This was a real bummer when I went to use roo on larger codebases that have .gits throughout. 

### Test Procedure
run the following tests:
src/core/webview/__tests__/ClineProvider.spec.ts
src/services/checkpoints/__tests__/ShadowCheckpointService.spec.ts
src/services/file-changes/__tests__/FileChangeManager.test.ts
webview-ui/src/components/file-changes/__tests__/FilesChangedOverview.spec.tsx
webview-ui/src/components/settings/__tests__/UISettings.spec.tsx
webview-ui/src/context/__tests__/ExtensionStateContext.spec.tsx
check all languages exist. 

open chat and ask it to modify some files, make new files, different types of tool calls, etc. accept and reject the changes
open settings menu and under interfaces, turn off the feature, go through the steps again 
turn on the feature on last time and go through the steps again. 

make sure checkpoints still work. 
when restoring a checkpoint, fco clears and disappears. 

### Pre-Submission Checklist

- [X] **Issue Linked**: This PR is linked to an approved GitHub Issue (see "Related GitHub Issue" above).
- [X] **Scope**: My changes are focused on the linked issue (one major feature/fix per PR). <- actually true now
- [X] **Self-Review**: I have performed a thorough self-review of my code.
- [X] **Testing**: New and/or updated tests have been added to cover my changes (if applicable).
- [X] **Documentation Impact**: I have considered if my changes require documentation updates (see "Documentation Updates" section below).
- [X] **Contribution Guidelines**: I have read and agree to the [Contributor Guidelines](/CONTRIBUTING.md).

#### Before:
image

#### After Checkpoint is hit: 
image
image
image

#### Settings option: 
image

### Documentation Updates

You can add a line like, "After every checkpoint, Files changed will appear in chat. You have the option to accept or reject them on a file by file basis. to turn this feature off, find it in the settings menu" but I don't think it makes much to consider it significant. 

### Additional Notes

I consolidated the feature, made a new branch, pushed just feature changes, and squashed it into a single commit to only contain changes made that involve this feature. 

----

> [!IMPORTANT]
> Introduces a feature to track and manage file changes between checkpoints, with UI updates, backend integration, and internationalization support.
> 
>   - **Feature**:
>     - Adds `FilesChangedOverview` component to display modified files between checkpoints in `TaskHeader.tsx`.
>     - Implements `FCOMessageHandler` in `FCOMessageHandler.ts` to handle file change messages.
>     - Introduces `FileChangeManager` in `FileChangeManager.ts` for managing file changes.
>   - **Backend**:
>     - Updates `checkpointRestore` and `getCheckpointService` in `index.ts` to integrate file change tracking.
>     - Modifies `applyDiffToolLegacy`, `attemptCompletionTool`, `insertContentTool`, and `searchAndReplaceTool` to track file edits.
>   - **Tests**:
>     - Adds tests for `FCOMessageHandler` and `FileChangeManager` in `FCOMessageHandler.test.ts` and `FileChangeManager.test.ts`.
>   - **Internationalization**:
>     - Updates `settings.json` for Italian, Russian, and Chinese locales to include new feature descriptions.
>   - **Misc**:
>     - Adds `useDebouncedAction` hook in `useDebouncedAction.ts` for debouncing actions.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=RooCodeInc%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for 5f7999d55fe5c9a1d192a0dc091b4ec21b5875dc. You can [customize](https://app.ellipsis.dev/RooCodeInc/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: src/services/file-changes/FCOMessageHandler.ts [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+		this.provider.postMessageToWebview({
+			type: "filesChanged",
+			filesChanged: undefined,
+		})
+	}
+
+	/**
+	 * Set up checkpoint event listener for universal baseline management
+	 */
+	private setupCheckpointListener(task: any): void {
```
Comment (ghost) [inline src/services/file-changes/FCOMessageHandler.ts]: Memory leak risk: The checkpoint event listener setup in setupCheckpointListener() may not be properly cleaned up in all scenarios. Consider ensuring removeCheckpointListener() is called in all cleanup paths, including when the provider is disposed or when switching between tasks.

File: src/services/file-changes/FCOMessageHandler.ts [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+		relativeFilePath: string,
+		fromCheckpoint: string,
+		checkpointService: any,
+	): Promise<void> {
+		if (!checkpointService?.restoreFileFromCheckpoint) {
+			throw new Error("Checkpoint service does not support per-file restore")
+		}
+
+		try {
+			await checkpointService.restoreFileFromCheckpoint(fromCheckpoint, relativeFilePath)
```
Comment (ghost) [inline src/services/file-changes/FCOMessageHandler.ts:598]: Error handling concern: The revertFileToCheckpoint method doesn't handle all error cases properly. If restoreFileFromCheckpoint throws an error other than 'file not found', the workspace could be left in an inconsistent state. Consider adding more robust error recovery.

File: webview-ui/src/components/file-changes/FilesChangedOverview.tsx [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+
+	// Self-managed state
+	const [changeset, setChangeset] = React.useState<FileChangeset | null>(null)
+	const [isInitialized, setIsInitialized] = React.useState(false)
+
+	const files = React.useMemo(() => changeset?.files || [], [changeset?.files])
+	const [isCollapsed, setIsCollapsed] = React.useState(true)
+
+	// Performance optimization: Use virtualization for large file lists
+	const VIRTUALIZATION_THRESHOLD = 50
```
Comment (ghost) [inline webview-ui/src/components/file-changes/FilesChangedOverview.tsx:38]: Performance consideration: The virtualization threshold of 50 files might be too high for slower machines. Consider lowering it to 20-30 files. This would improve performance on lower-end devices while still handling most common use cases efficiently.

File: src/services/file-changes/FileChangeManager.ts [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+	 * Get a specific file change
+	 */
+	public getFileChange(uri: string): FileChange | undefined {
+		return this.changeset.files.find((file) => file.uri === uri)
+	}
+
+	/**
+	 * Accept a specific file change
+	 */
+	public async acceptChange(uri: string): Promise<void> {
```
Comment (ghost) [inline src/services/file-changes/FileChangeManager.ts:88]: Potential race condition: Multiple concurrent accept/reject operations could cause race conditions when modifying acceptedBaselines and changeset.files. Consider implementing a queue or mutex pattern to serialize these operations.

**Comment by daniel-lxs** [comment]
Hey @playcations can you take a look at the failing unit test?

**Comment by playcations** [comment]
It is missing tests I have at home that I missed pushing. I will get those in when I get home in about 4 hours.

**Comment by playcations** [comment]
Added tests and updated language files. fixed found edge cases.

**Comment by playcations** [comment]
I will be closing this branch as I am working on a refactor with better separation of concerns, less dependencies on tools, etc. Hopefully will be ready later this week.
